source/material_model/averaging.cc

```diff
@@ -0,0 +1,412 @@
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    // Do the requested averaging operation for one array.
+    template <int dim>
+    void
+    Averaging<dim>::average (const std::string &averaging_operation,
+                             const std::vector<Point<dim> >    &position,
+                             std::vector<double>           &values_out) const
+    {
+      // if an output field has not been filled (because it was
+      // not requested), then simply do nothing -- no harm no foul
+      if (values_out.size() == 0)
+        return;
+
+      const unsigned int N = values_out.size();
+      std::string averaging_operation_lower = averaging_operation;
+
+      // switch to lower case, so the input becomes independent of case
+      std::transform(averaging_operation_lower.begin(), averaging_operation_lower.end(), averaging_operation_lower.begin(), ::tolower);
+
+      // perform the requested averaging
+      if (averaging_operation_lower == "none")
+        {
+        }
+      else if (averaging_operation_lower == "arithmetic average")
+        {
+          double sum = 0;
+          for (unsigned int i=0; i<N; ++i)
+            sum += values_out[i];
+
+          const double average = sum/N;
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = average;
+        }
+      else if (averaging_operation_lower == "harmonic average")
+        {
+          double sum = 0;
+          for (unsigned int i=0; i<N; ++i)
+            sum += 1./values_out[i];
+
+          const double average = 1./(sum/N);
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = average;
+        }
+      else if (averaging_operation_lower == "geometric average")
+        {
+          double prod = 1;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              Assert (values_out[i] >= 0,
+                      ExcMessage ("Computing the geometric average "
+                                  "only makes sense for non-negative "
+                                  "quantities."));
+              prod *= values_out[i];
+            }
+
+          const double average = std::pow (prod, 1./N);
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = average;
+        }
+      else if (averaging_operation_lower ==  "pick largest")
+        {
+          double max = -std::numeric_limits<double>::max();
+          for (unsigned int i=0; i<N; ++i)
+            max = std::max(max, values_out[i]);
+
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = max;
+          //break;
+        }
+      else if (averaging_operation_lower == "nwd arithmetic average")
+        {
+          double sum_value = 0;
+          double sum_weights = 0;
+          std::vector<double> temp_values(N,0);
+          double max_distance = 0;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
+              double h = bell_shape_limit;
```

Comment:
Unless you plan to update `h`, just use the original variable to avoid confusion.